### PR TITLE
[Discussion] main loop: mainthread vs. simthread updates.

### PR DESCRIPTION
**Pull request by only-a-ptr**
*Discussion PR, do not merge yet!*

While working on [LuaPowertrain](http://www.rigsofrods.com/threads/120811-Dev-LuaPowertrain-project), I got blocked out by RoR's architecture - the engine/brakes/otherstuff are updated from both mainthread (gameplay/LandVehicleSimulation.cpp) and simthread (gameplay/BeamEngine.cpp) without proper synchronization. By coincidence, it always worked fine, but it can't stay that way. For one thing, it blows up when any scripting is added.

I refactored the code towards this architecture:
 * Mainthread reads and buffers all user inputs. Also updates GUI+visuals from vehicle state snapshot.
 * Simthread reads the input buffer and performs updates. At the end, it takes a snapshot of vehicle state.
 * [sync] Input buffer is passed to simthread, snapshot is passed to mainthread.

Now there's a question: what next?
 * Stabilize the code as it is and implement Lua scripting of the powertrain.
 * Continue the redesign and make simthread exclusively responsible for ALL vehicle updates. Powertrain took care of most, but there's also steering, commands, hooks...

@Speciesx I need you to build and test this. I think the vehicles behave differently.
@ulteq I need you to review this.

Branch: https://github.com/only-a-ptr/rigs-of-rods/tree/simulation-powertrain-MT-rev2

### CODE REVIEW

File: source/main/gameplay/Powertrain.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+/** 
+    @file
+    @date   01/2016
+    @author Petr Ohlidal
+*/
+
+#pragma once
+
+#include "ForwardDeclarations.h"
+#include "transmission.h"
```
Comment (Speciesx) [inline source/main/gameplay/Powertrain.h]: #include "datatypes/transmission.h"

shouldn't be this?

Comment (only-a-ptr) [inline source/main/gameplay/Powertrain.h]: @Speciesx Could be, but it's a detail.

**Comment by Speciesx** [comment]
I've noticed so far:
- In automatic shift gearbox vehicles start without starter, switch on ignition and press accelerator 
- Arcade controls are active when start the engine in non automatic shift gearbox
- Switch between 4k and 8k Speedometer doesn't work

**Comment by only-a-ptr** [comment]
NOTE TO ALL:
Do not merge this. It's only here to attract discussion (it's not working, anyway...)

When we decide what to do about it next, I'll create a new branch from current upstream and rebase+polish the commits.

**Comment by only-a-ptr** [comment]
Decided in dev chat:
now that there's been so much progress lately based around the current upstream, I agree with @mikadou : Let's first stabilize upstream, make a release and then start restructuring hard.

>  There exist better means for navigating code
@mikadou Show me.

**Comment by mikadou** [comment]
@only-a-ptr I'm using rtags on Linux. I might be wrong, but I'm pretty sure  visualstudio also has some kind of 'find usages' feature built-in

**Comment by only-a-ptr** [comment]
@mikadou VisualStudio's "find all references" sucks - it's very slow and it's unreliable - it's practically just a dumb fulltext searcher (it shows all symbols of the same name) but some results get left out. Unusable.

I agree the repetitive function names aren't elegant, but they're so damn useful! Any other way of reference searching is technically super-complex and requires extra utilities. Personally I think the very reason for programming languages still being text-based is text indexing and plaintext searching. In the company I work for, the most valuable tool is [OpenGrok](https://opengrok.github.io/OpenGrok/) server.

**Comment by mikadou** [comment]
I see. I wasn't aware that VisualStudio's search is not that sophisticated. Thanks for telling me about the OpenGrok tool. Looks very interesting. I'm curious, does OpenGrok provide context-aware search? Or is it also basically also a dumb fulltext searcher?

**Comment by only-a-ptr** [comment]
@mikadou OpenGrok understands code and uses the data to cross-link the documets and to annotate results, but not for searching itself.
These are supported search fields: 
Full Search - fulltext	
Definition - find struct/variable/function definition by symbol name
Symbol 	- basically case-sensitive fulltext
File Path 	- filters results
History 	- dunno what this does

**Pull request closed without merging**